pkg/datasource/getInsights.go

```diff
@@ -0,0 +1,101 @@
+package datasource
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	xray "github.com/grafana/x-ray-datasource/pkg/xray"
+)
+
+type GetInsightsQueryData struct {
+	State string `json:"state"`
+	Group string `json:"group"`
+}
+
+func (ds *Datasource) getInsights(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	xrayClient, err := ds.xrayClientFactory(&req.PluginContext)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &backend.QueryDataResponse{
+		Responses: make(map[string]backend.DataResponse),
+	}
+
+	for _, query := range req.Queries {
+		response.Responses[query.RefID] = getSingleInsight(xrayClient, query)
+	}
+
+	return response, nil
+}
+
+func getSingleInsight(xrayClient XrayClient, query backend.DataQuery) backend.DataResponse {
+	queryData := &GetInsightsQueryData{}
+	err := json.Unmarshal(query.JSON, queryData)
+
+	if err != nil {
+		return backend.DataResponse{
+			Error: err,
+		}
+	}
+
+	var states []string = []string{strings.ToUpper(queryData.State)}
+
+	log.DefaultLogger.Debug("states", "states", states)
+
+	if queryData.State == "All" || len(queryData.State) == 0 {
+		states = nil
+	}
+
+	insightsResponse, err := xrayClient.GetInsightSummaries(&xray.GetInsightSummariesInput{
+		StartTime: &query.TimeRange.From,
+		EndTime:   &query.TimeRange.To,
+		States:    aws.StringSlice(states),
```

Comment:
What is the difference between String and StringSlice?

Reply:
It is the same as String but for String arrays.